### PR TITLE
Make gce/upgrade.sh script less sensitive

### LINKED ISSUE #37715: upgrade.sh does not change node version properly
See
https://groups.google.com/d/msg/kubernetes-dev/C8s1boeUCZ4/gju__NuJCgAJ
and subsequent messages in that thread.

### PR DESCRIPTION
**Pull request by gmarek**
Fix #37715

This should fix the upgrade script. It's after midnight, so I don't have resources to test it thoroughly, but I tried it once and it seems to work. Do we have any bash wizard to review?

cc @davidopp @roberthbailey @mtaufen @bprashanth

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37724)

**Comment by bprashanth** [comment]
This is kind of the direction I was going in on the issue, but I think we decided to leave it as is

**Comment by gmarek** [comment]
@k8s-bot gci gce test this issue: ##38083

File: cluster/gce/upgrade.sh [hunk 1]
```diff
@@ -328,35 +328,45 @@ node_prereqs=false
 local_binaries=false
 env_os_distro=false
 
-while getopts ":MNPlho" opt; do
-  case ${opt} in
-    M)
+args=$(getopt -o MNPlho -- "$@")
```
Comment (roberthbailey) [inline cluster/gce/upgrade.sh]: This doesn't work on my mac. If I echo $args right after this line when running `upgrade.sh -h` I see `-- MNPlho -- -h` and the first iteration through the while loop has `$1` set to `--`. And no arguments are parsed.

File: cluster/gce/upgrade.sh [hunk 2]
```diff
@@ -328,35 +328,50 @@ node_prereqs=false
 local_binaries=false
 env_os_distro=false
 
-while getopts ":MNPlho" opt; do
-  case ${opt} in
-    M)
+if [[ -n $(echo $OSTYPE | grep "darwin") ]]; then
+  # OS X version
+  args=$(getopt MNPlho $@)
```
Comment (zmerlynn) [inline cluster/gce/upgrade.sh]: Yeah, this is the portable version. I spent a while looking when I first wrote the script.

I don't like this OS-level check, it makes the script behave differently on OS X than Linux for little purpose?

Comment (gmarek) [inline cluster/gce/upgrade.sh]: Right. Fixed. @zmerlynn PTAL.

**Comment by roberthbailey** [comment]
This arg parsing worked for me, but still doesn't deal well with the issue where a flag comes after a non-flag (the flag is still ignored). So is it an improvement over what we currently have?

**Comment by roberthbailey** [comment]
I just extracted the arg parsing into a short script to see how it worked:

```shell
$ cat args.sh 
#!/bin/bash

master_upgrade=true
node_upgrade=true
node_prereqs=false
local_binaries=false
env_os_distro=false

args=$(getopt MNPlho $@)
if [[ $? -ne 0 ]]; then
  >&2 echo "Failed to parse options"
  exit 1
fi
eval set -- "${args}"

while true; do
  case $1 in
    -M)
      shift
      node_upgrade=false
      ;;
    -N)
      shift
      master_upgrade=false
      ;;
    -P)
      shift
      node_prereqs=true
      ;;
    -l)
      shift
      local_binaries=true
      ;;
    -o)
      shift
      env_os_distro=true
      ;;
    -h)
      usage
      exit 0
      ;;
    --)
      shift
      break;
      ;;
  esac
done

echo "$master_upgrade"
echo "$node_upgrade"
echo "$node_prereqs"
echo "$local_binaries"
echo "$env_os_distro"

echo "args: $@"
```

What does that produce on linux when you run

```shell
$ ./args.sh -M foo bar
true
false
false
false
false
args: foo bar
```
and

```shell
$ ./args.sh foo bar -M
true
true
false
false
false
args: foo bar -M
```

You can see in the second case that the master_upgrade flag isn't changed, and the `-M` isn't found by the arg processing.

**Comment by roberthbailey** [comment]
I think just making sure we don't have any dangling flags (e.g. https://github.com/kubernetes/kubernetes/pull/37723) is a safer solution.

**Comment by gmarek** [comment]
This is a problem with a 'portable' version. Original one (that doesn't work on OS X) does:
```shell
go/src/k8s.io/kubernetes$ cat test.sh 
#!/bin/bash

master_upgrade=true
node_upgrade=true
node_prereqs=false
local_binaries=false
env_os_distro=false

args=$(getopt -o MNPlho -- $@)
if [[ $? -ne 0 ]]; then
  >&2 echo "Failed to parse options"
  exit 1
fi
eval set -- "${args}"

while true; do
  case $1 in
    -M)
      shift
      node_upgrade=false
      ;;
    -N)
      shift
      master_upgrade=false
      ;;
    -P)
      shift
      node_prereqs=true
      ;;
    -l)
      shift
      local_binaries=true
      ;;
    -o)
      shift
      env_os_distro=true
      ;;
    -h)
      usage
      exit 0
      ;;
    --)
      shift
      break;
      ;;
  esac
done

echo "$master_upgrade"
echo "$node_upgrade"
echo "$node_prereqs"
echo "$local_binaries"
echo "$env_os_distro"

echo "args: $@"
go/src/k8s.io/kubernetes$ ./test.sh foo bar -M
true
false
false
false
false
args: foo bar
go/src/k8s.io/kubernetes$ ./test.sh -M foo bar
true
false
false
false
false
args: foo bar
```

Maybe we should go back to the OS-guard for this?

**Comment by roberthbailey** [comment]
I guess my question is whether #37723 is sufficient or if we care about fancier flag parsing?

**Comment by gmarek** [comment]
The answer depends on what's the goal for this tool. If it's supposed to be production-level one which we'll be keeping up to date (which may mean adding more parameters) I'd say we want fancier flag parsing. Otherwise it's may not worth the effort.

**Comment by zmerlynn** [comment]
@gmarek: This script was never intended to be production-grade, and still has an `!!! EXPERIMENTAL !!!` warning in the top-comment and usage.

**Pull request closed without merging**